### ISSUE TITLE
Question: 'Do Nothing'

### ISSUE THREAD

**Issue by kmrsh** [kaitai-io/kaitai_struct#861]
I have similar situation as described in [Do Nothing](https://doc.kaitai.io/user_guide.html#do-nothing) in the documentation.

~~~yml
seq:
  - id: rec_type
    type: u4
  - id: body
    type:
      switch-on: rec_type
      cases:
        1: rec_type_1
        2: rec_type_2
        3: dummy # <= must ignore rec_type=3!
        _: rec_type_others
~~~

where I want to have the `body` of given type only if the `rec_type` is of certain cases only. all others should be neglected
there are hundreds of records in a single file, Something similar to below

~~~yml
seq:
  - id: event
    type: t_event_type
    repeat: eos

types:
  t_event_type:
    seq:
      - id: event_length
        type: u2
      - id: event_type
        type: u2
      - id: event_data                              <--- this is where body comes with certain event types, ignoring other types
        size: event_length - 7
~~~

now if I introduce `siwtch-on` like above and defining the types for bodies of only records I needed.
similar to the documentation, how does kaitai know how many bytes to ignore to goto the next record? I understand how kaitai know when to start next record if we have given the data type in a squence, but if we define dummy_x types as below, how does it know the ignored event's body is over and parse the next event?

~~~yml
ypes:
  # One can use empty JSON object syntax to avoid specifying any of
  # `seq`, `instances`, etc, sections.
  dummy_1: {}
  # One can use explicit doc to note that there's nothing there.
  dummy_2:
    doc: This type is intentionally left blank.
  # One can use empty `seq` or `instances` or `types` section, any
  # other empty sections, or any combination of thereof.
  dummy_3:
    seq: []
    instances: {}
    types: {}
  # One can use a very explicit notion of the fact that we want to parse 0 bytes.
  dummy_4:
    seq:
      - id: no_value
        size: 0
~~~

**Comment by KOLANICH** [comment]
>now if I introduce siwtch-on like above and defining the types for bodies of only records I needed.
similar to the documentation, how does kaitai know how many bytes to ignore to goto the next record?

1. You have specified `size`.
2. You should really practice a bit by compiling different specs and looking at the generated code.

**Comment by kmrsh** [comment]
Yes in this case I have specified the size, but my question is about "Do nothing" section of the documentation, where it doesn't specify any size for unwanted dummy types. that's my question?

I'll paste the sample here (https://doc.kaitai.io/user_guide.html#do-nothing)

~~~yml
  - id: rec_type
    type: u4
  - id: body
    type:
      switch-on: rec_type
      cases:
        1: rec_type_1
        2: rec_type_2
        3: dummy # <= must ignore rec_type=3!
        _: rec_type_others
~~~

~~~yml
types:
  # One can use empty JSON object syntax to avoid specifying any of
  # `seq`, `instances`, etc, sections.
  dummy_1: {}
  # One can use explicit doc to note that there's nothing there.
  dummy_2:
    doc: This type is intentionally left blank.
  # One can use empty `seq` or `instances` or `types` section, any
  # other empty sections, or any combination of thereof.
  dummy_3:
    seq: []
    instances: {}
    types: {}
  # One can use a very explicit notion of the fact that we want to parse 0 bytes.
  dummy_4:
    seq:
      - id: no_value
        size: 0
~~~

Now forget about my scenario and considering the example given in the doc, how does kaitai ignore ` 3: dummy # <= must ignore rec_type=3!` (or dummy_1 or dummy_2 etc) and move on to next record of dummy_x types are without any size given?

**Comment by KOLANICH** [comment]
There is no magic dum y type in KS. So, it is not really ignored. So an object is created, that consumes memory, but doesn't consume any bytes from stream. It may be beneficial in some cases, but only for duck-typed languages currently. If `body` has no `size`, it is better to just omit the case, this way the element will be `null`, `None`, `nullptr` and other language built-in empty value.

**Comment by kmrsh** [comment]
I don't think I'm clear enough or you don't understand what I'm saying,

what Im saying is

~~~yml
...
3: dummy # <= must ignore rec_type=3!
...
~~~

here you ignore certain `rec_type`. ignoring doesn't mean actual binary file doesn't have any bytes, for that record, it has but we don't care about that type. then how does kaitai knows how to ignore these bytes and jump to the next record of type which we're interested.

**Comment by generalmimon** [comment]
> ignoring doesn't mean actual binary file doesn't have any bytes, for that record, it has, but we don't care about that type. then **how does kaitai know** how to ignore these bytes and jump to the next record of type which we're interested.

It doesn't know. There is no magic going on there - if `seq` of any type (i.e. of the type `dummy`) is empty, it will simply parse 0 bytes (unless you put it into a substream with fixed size). The section [Do nothing](https://doc.kaitai.io/user_guide.html#do-nothing) in the documentation tries to express that some special record types (`rec_type = 3`) have an empty body/payload (0 bytes).

If you only _don't care_ about the record contents but still want to advance the stream position, you **need** to parse the record data anyway (unless you want to play with parse `instances` and `pos`ition the next record onto a byte offset calculated by an arithmetic expression, but I don't recommend that - it's too complicated for little effect). Just put a fixed-sized byte array in `rec_type_3`, this is the best thing you can do:

```ksy
types:
  rec_type_3:
    seq:
      - id: data
        size: 32
```

**Comment by kmrsh** [comment]
> > ignoring doesn't mean actual binary file doesn't have any bytes, for that record, it has, but we don't care about that type. then **how does kaitai know** how to ignore these bytes and jump to the next record of type which we're interested.
> 
> It doesn't know. There is no magic going on there - if `seq` of any type (i.e. of the type `dummy`) is empty, it will simply parse 0 bytes (unless you put it into a substream with fixed size). The section [Do nothing](https://doc.kaitai.io/user_guide.html#do-nothing) in the documentation tries to express that some special record types (`rec_type = 3`) have an empty body/payload (0 bytes).
> 
> If you only _don't care_ about the record contents but still want to advance the stream position, you **need** to parse the record data anyway (unless you want to play with parse `instances` and `pos`ition the next record onto a byte offset calculated by an arithmetic expression, but I don't recommend that - it's too complicated for little effect). Just put a fixed-sized byte array in `rec_type_3`, this is the best thing you can do:
> 
> ```yaml
> types:
>   rec_type_3:
>     seq:
>       - id: data
>         size: 32
> ```

Yes, that's what I thought, I was wondering why there are certain record types in a file while no bytes (to represent that record), Your answer makes sense

I would like to see some kaitai commands that we can use, to ignore storing certain byte array completely if we don't need them. for some large binary files we process, we only care about certain record types, and we need map them to structures, but we don't need to waste memory keeping some byte arrays of many kilo-mega bytes which we're not interested at all. This makes use of the memory efficient when it comes large files in high velocity processing. 

Do you know such kaitai construct that we can say, don't even keep byte array if it encounters certain record type. One option I could think is to use [Custom processors](https://doc.kaitai.io/user_guide.html#custom-process), where we simply return empty array. But I was looking whether Kaitai already has something that can be used
